From f85342ca80da67c6a7553a7248e33aced23b44ee Mon Sep 17 00:00:00 2001
From: Keqian Zhu <zhukeqian1@huawei.com>
Date: Tue, 8 Mar 2022 22:06:05 +0800
Subject: [PATCH] vfio/iommu_type1: Fix the logic of updating
 num_non_hwdbm_domains
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: f85342ca80da67c6a7553a7248e33aced23b44ee
Modified-by-SEL: No


virt inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I4WK5B
CVE: NA

------------------------------

When a newly created iommu_domain is not HWDBM capable, the vfio_iommu's
num_non_hwdbm_domains field should increase but failed to increase. This
fixes this bug.

Fixes: d2373c56459f (vfio/iommu_type1: Add HWDBM status maintenance)
Signed-off-by: Keqian Zhu <zhukeqian1@huawei.com>
Tested-by: Kunkun Jiang <jiangkunkun@huawei.com>
Reviewed-by: Kunkun Jiang <jiangkunkun@huawei.com>
Reviewed-by: Zenghui Yu <yuzenghui@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 drivers/vfio/vfio_iommu_type1.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/vfio/vfio_iommu_type1.c b/drivers/vfio/vfio_iommu_type1.c
index 5daceec48811..f556b572c86d 100644
--- a/drivers/vfio/vfio_iommu_type1.c
+++ b/drivers/vfio/vfio_iommu_type1.c
@@ -2420,7 +2420,7 @@ static void vfio_iommu_update_hwdbm(struct vfio_iommu *iommu,
 	bool num_non_hwdbm_zeroed = false;
 	bool log_enabled, should_enable;
 
-	if (old_hwdbm && !new_hwdbm && attach) {
+	if ((old_hwdbm || singular) && !new_hwdbm && attach) {
 		iommu->num_non_hwdbm_domains++;
 	} else if (!old_hwdbm && new_hwdbm && !attach) {
 		iommu->num_non_hwdbm_domains--;
-- 
2.26.2

